pkg/types/types.go

```diff
@@ -27,28 +27,74 @@ func NewRetrievalCandidate(pid peer.ID, rootCid cid.Cid, protocols ...metadata.P
 	}
 }
 
+// retrieval task is any task that can be run to produce a result
+type RetrievalTask interface {
+	Run() (*RetrievalStats, error)
+}
+
 type Retriever interface {
 	Retrieve(ctx context.Context, request RetrievalRequest, events func(RetrievalEvent)) (*RetrievalStats, error)
 }
 
 type CandidateFinder interface {
-	FindCandidates(ctx context.Context, request RetrievalRequest, events func(RetrievalEvent)) ([]RetrievalCandidate, error)
+	FindCandidates(ctx context.Context, request RetrievalRequest, events func(RetrievalEvent), onCandidates func([]RetrievalCandidate)) error
+}
+
+func MakeAsyncCandidates(buffer int) (InboundAsyncCandidates, OutboundAsyncCandidates) {
+	asyncCandidates := make(chan []RetrievalCandidate, buffer)
+	return asyncCandidates, asyncCandidates
+}
+
+type InboundAsyncCandidates <-chan []RetrievalCandidate
+
+func (ias InboundAsyncCandidates) Next(ctx context.Context) (bool, []RetrievalCandidate, error) {
+	// prioritize cancelled context
+	select {
+	case <-ctx.Done():
+		return false, nil, ctx.Err()
+	default:
+	}
+	select {
+	case <-ctx.Done():
+		return false, nil, ctx.Err()
+	case next, ok := <-ias:
+		return ok, next, nil
+	}
 }
 
-type CandidateRetrieval interface {
-	RetrieveFromCandidates([]RetrievalCandidate) (*RetrievalStats, error)
+type OutboundAsyncCandidates chan<- []RetrievalCandidate
+
+func (oas OutboundAsyncCandidates) SendNext(ctx context.Context, next []RetrievalCandidate) error {
+	select {
+	case <-ctx.Done():
+		return ctx.Err()
+	case oas <- next:
+		return nil
+	}
+}
+
+type CandidateRetreival interface {
```

Comment:
```suggestion
type CandidateRetrieval interface {
```

Reply:
gonna do this locally to get the accross the board rename through